stdlib/iarray.ml

```diff
@@ -0,0 +1,95 @@
+(* CR aspectorzabusky: This needs a copyright header; should I copy [array.ml]? *)
+
+open! Stdlib
+
+[@@@ocaml.flambda_o3]
+
+(* An alias for the type of immutable arrays. *)
+type +'a t = 'a iarray
+
+(* Array operations *)
+
+external length       : 'a iarray -> int       = "%array_length"
+external get          : 'a iarray -> int -> 'a = "%array_safe_get"
+external ( .:() )     : 'a iarray -> int -> 'a = "%array_safe_get"
+external unsafe_get   : 'a iarray -> int -> 'a = "%array_unsafe_get"
+
+(* Immutable and mutable arrays have the same runtime representation; we
+   construct immutable arrays by constructing mutable arrays and then blindly
+   casting them to become immutable.  This is safe here because:
+   1. None of these functions mutate their array inputs;
+   2. None of these functions hold on to their array inputs; and
+   3. All of these functions return fresh arrays if they return an array. *)
+let init =
```

Comment:
Personally I'd rather avoid uses of `Obj.magic` here -- there are very few uses in the stdlib, which seems like a good thing.  Can we not just copy the code of the original `Array` module and edit it as required (in which case the INRIA copyright notice must be preserved)?